### PR TITLE
Send some ephemeral events to appservices

### PR DESCRIPTION
**Pull request by Half-Shot**
This PR is split from #8366.

This change will cause Synapse to send ephemeral events to appservices in accordance with [MSC2409](https://github.com/matrix-org/matrix-doc/pull/2409). This PR only sends typing, presence and read receipt information to appservices, but does not handle device information such as to_device messages or device list changes which will be part of #8366.

### CODE REVIEW

**Comment by Half-Shot** [comment]
I have started at this for a long time, but unable to understand why `test_multiple_service_queues` and `test_send_single_event_with_queue` are the way they are.

It feels like it's a racy test, as we enqueue two events for the appservice which will spawn a background task to process both of them, and then we will also check that `self.txn_ctrl.send.assert_called_with` handled the *first* event.

When I run the test against develop it obviously works, only reporting that it processed the first event. When I run it on my branch, it processes the first event and the second event. I'd expect the latter behavior to be correct.

**Comment by anoadragon453** [comment]
@Half-Shot I agree with you on `test_multiple_service_queues` in that it probably should be processing both events in a single transaction? But `test_send_single_event_with_queue` seems to behave as you describe and yet it still fails.

Needs more investigation.

**Comment by Half-Shot** [comment]
Now that that's out of the way, I'm going to write some tests :)

File: changelog.d/8437.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Implement MSC2409 to send typing, read receipts and presence events to appservices.
```
Comment (clokep) [inline changelog.d/8437.feature:1]: Please link to the MSC. 👍

File: synapse/appservice/__init__.py [hunk 2]
```diff
@@ -32,38 +32,6 @@ class ApplicationServiceState:
     UP = "up"
 
 
-class AppServiceTransaction:
```
Comment (clokep) [inline synapse/appservice/__init__.py:35]: Moving this makes the diff rather difficult to follow. It looks like the difference is the `__init__` takes an `ephemeral` and pipes it into `push_bulk`?

File: synapse/appservice/__init__.py [hunk 3]
```diff
@@ -188,14 +158,14 @@ async def _matches_user(self, event, store):
...
         return does_match
 
     @cached(num_args=1, cache_context=True)
-    async def _matches_user_in_member_list(self, room_id, store, cache_context):
-        member_list = await store.get_users_in_room(
-            room_id, on_invalidate=cache_context.invalidate
-        )
+    async def matches_user_in_member_list(
+        self, room_id: str, store, cache_context: _CacheContext
+    ):
```
Comment (clokep) [inline synapse/appservice/__init__.py:167]: Please add a type for `store` and a return type (`bool` it looks like).

It looks like the cache context is no longer being used, should the `cached` decorator be updated to stop providing it?

Since we're making this "public", please add a docstring.

File: synapse/appservice/__init__.py [hunk 4]
```diff
@@ -239,6 +209,21 @@ async def is_interested(self, event, store=None) -> bool:
 
         return False
 
+    @cached(num_args=1, cache_context=True)
+    async def is_interested_in_presence(
+        self, user_id: UserID, store, cache_context: _CacheContext
+    ):
```
Comment (clokep) [inline synapse/appservice/__init__.py:215]: Please add a docstring! (And generally the same comments as for `matches_user_in_member_list`.)

File: synapse/appservice/api.py [hunk 5]
```diff
@@ -201,7 +202,13 @@ async def _get() -> Optional[JsonDict]:
         key = (service.id, protocol)
         return await self.protocol_meta_cache.wrap(key, _get)
 
-    async def push_bulk(self, service, events, txn_id=None):
+    async def push_bulk(
```
Comment (clokep) [inline synapse/appservice/api.py:205]: While we're here, can you add a docstring here? IN particular what is `ephemeral` supposed to be? I'm guessing it is actually `JsonDict`, not `Any`?

File: synapse/appservice/scheduler.py [hunk 6]
```diff
@@ -114,7 +122,19 @@ def enqueue(self, service, event):
...
+    def enqueue_ephemeral(self, service: ApplicationService, events: List[Any]):
+        self.queued_ephemeral.setdefault(service.id, []).extend(events)
+
+        # start a sender for this appservice if we don't already have one
+
+        if service.id in self.requests_in_flight:
+            return
+
+        run_as_background_process(
+            "as-sender-%s" % (service.id,), self._send_request, service
```
Comment (clokep) [inline synapse/appservice/scheduler.py:134]: Should this differ from the name used in `enqueue`?

File: synapse/appservice/scheduler.py [hunk 7]
```diff
@@ -204,7 +232,7 @@ def start_recoverer(self, service):
         recoverer.recover()
         logger.info("Now %i active recoverers", len(self.recoverers))
 
-    async def _is_service_up(self, service):
+    async def is_service_up(self, service: ApplicationService):
```
Comment (clokep) [inline synapse/appservice/scheduler.py:235]: I don't see why we are renaming this here? Please add a return type.

File: synapse/handlers/appservice.py [hunk 8]
```diff
@@ -158,6 +163,90 @@ async def handle_room_events(events):
...
+        from_key = await self.store.get_type_stream_id_for_appservice(
+            service, "read_receipt"
+        )
+        receipts_source = self.event_sources.sources["receipt"]
+        receipts, _ = await receipts_source.get_new_events_as(
+            service=service, from_key=from_key
+        )
+        return receipts
+
+    async def _handle_as_presence(self, service: ApplicationService, users: List[str]):
```
Comment (clokep) [inline synapse/handlers/appservice.py:224]: Is the "as" here appservice or the word "as"? (I'm confused why this is using a different naming scheme from above).

File: synapse/handlers/typing.py [hunk 9]
```diff
@@ -430,6 +431,27 @@ def _make_event_for(self, room_id):
...
+        self, from_key: int, service: ApplicationService, **kwargs
+    ):
+        with Measure(self.clock, "typing.get_new_events_as"):
+            from_key = int(from_key)
+            handler = self.get_typing_handler()
+
+            events = []
+            for room_id in handler._room_serials.keys():
+                if handler._room_serials[room_id] <= from_key:
+                    print("Key too old")
```
Comment (clokep) [inline synapse/handlers/typing.py:444]: Left-over debugging code.

File: synapse/handlers/typing.py [hunk 10]
```diff
@@ -430,6 +431,27 @@ def _make_event_for(self, room_id):
             "content": {"user_ids": list(typing)},
         }
 
+    async def get_new_events_as(
+        self, from_key: int, service: ApplicationService, **kwargs
```
Comment (clokep) [inline synapse/handlers/typing.py:435]: Doesn't look like these new methods need to accept `kwargs`?

Can you add docstrings to them? 😄

File: synapse/handlers/appservice.py [hunk 11]
```diff
@@ -158,6 +163,90 @@ async def handle_room_events(events):
             finally:
                 self.is_processing = False
 
+    async def notify_interested_services_ephemeral(
```
Comment (clokep) [inline synapse/handlers/appservice.py:166]: Do we have any thoughts on how heavy this function might be? It looks like it might be called in the background though?

Can you add a docstring talking a bit more about how / when it runs?

File: synapse/storage/databases/main/appservice.py [hunk 12]
```diff
@@ -172,7 +174,12 @@ async def set_appservice_state(self, service, state) -> None:
             "application_services_state", {"as_id": service.id}, {"state": state}
         )
 
-    async def create_appservice_txn(self, service, events):
+    async def create_appservice_txn(
+        self,
+        service: ApplicationService,
+        events: List[EventBase],
```
Comment (clokep) [inline synapse/storage/databases/main/appservice.py:180]: Thanks for adding types here! Can you remove them from the docstring? And we should probably add the return type while we're at it (and remove it from below).

**Review by clokep — CHANGES_REQUESTED** [review]
Most of my comments can be summarized as "please add docstrings".

In general, could you add a bit more to the description about how / why?

File: synapse/appservice/__init__.py (hunk 2 above)
Comment (Half-Shot) [reply synapse/appservice/__init__.py:35]: It also takes an `ApplicationService`

File: synapse/appservice/scheduler.py (hunk 6 above)
Comment (Half-Shot) [reply synapse/appservice/scheduler.py:134]: I'm not sure it needs to? I've instead chosen to remove the duplicate code and make it one function.

**Comment by clokep** [comment]
@Half-Shot Looks like CI is unhappy? Might need to merge develop in?

**Comment by Half-Shot** [comment]
Looks like my type changes broke some tests, I'll take a look tomorrow.

File: synapse/handlers/appservice.py (hunk 11 above)
Comment (clokep) [reply synapse/handlers/appservice.py:166]: @Half-Shot Any idea on the initial question here?

File: synapse/handlers/appservice.py [hunk 13]
```diff
@@ -158,6 +163,101 @@ async def handle_room_events(events):
...
+                    if events:
+                        self.scheduler.submit_ephemeral_events_for_as(service, events)
+                    # We don't persist the token for typing_key for performance reasons
+                elif stream_key == "receipt_key":
+                    events = await self._handle_receipts(service)
+                    if events:
+                        self.scheduler.submit_ephemeral_events_for_as(service, events)
+                        await self.store.set_type_stream_id_for_appservice(
+                            service, "read_receipt", new_token
+                        )
```
Comment (clokep) [inline synapse/handlers/appservice.py:204]: Should this be in the `if events` clause or outside of it? (It does not match what is done for presence below.)

File: synapse/handlers/appservice.py [hunk 14]
```diff
@@ -158,6 +163,101 @@ async def handle_room_events(events):
...
+                    if events:
+                        self.scheduler.submit_ephemeral_events_for_as(service, events)
+                    await self.store.set_type_stream_id_for_appservice(
+                        service, "presence", new_token
+                    )
+
+    async def _handle_typing(self, service: ApplicationService, new_token: int):
+        typing_source = self.event_sources.sources["typing"]
+        # Get the typing events from just before current
+        typing, _key = await typing_source.get_new_events_as(
```
Comment (clokep) [inline synapse/handlers/appservice.py:216]:
```suggestion
        typing, _ = await typing_source.get_new_events_as(
```

File: synapse/handlers/appservice.py [hunk 15]
```diff
@@ -158,6 +163,101 @@ async def handle_room_events(events):
...
+        events = []
+        presence_source = self.event_sources.sources["presence"]
+        from_key = await self.store.get_type_stream_id_for_appservice(
+            service, "presence"
+        )
+        for user in users:
+            interested = await service.is_interested_in_presence(user, self.store)
+            if not interested:
+                continue
+            presence_events, _key = await presence_source.get_new_events(
```
Comment (clokep) [inline synapse/handlers/appservice.py:245]:
```suggestion
            presence_events, _ = await presence_source.get_new_events(
```

File: synapse/storage/databases/main/appservice.py [hunk 16]
```diff
@@ -172,15 +175,18 @@ async def set_appservice_state(self, service, state) -> None:
...
         """Atomically creates a new transaction for this application service
         with the given list of events.
 
         Args:
-            service(ApplicationService): The service who the transaction is for.
-            events(list<Event>): A list of events to put in the transaction.
-        Returns:
-            AppServiceTransaction: A new transaction.
+            service: The service who the transaction is for.
+            events: A list of events to put in the transaction.
```
Comment (clokep) [inline synapse/storage/databases/main/appservice.py:189]:
```suggestion
            events: A list of events to put in the transaction.

        Returns:
            A new transaction.
```

File: synapse/storage/databases/main/appservice.py [hunk 17]
```diff
@@ -172,15 +175,18 @@ async def set_appservice_state(self, service, state) -> None:
...
 
-    async def create_appservice_txn(self, service, events):
+    async def create_appservice_txn(
+        self,
+        service: ApplicationService,
+        events: List[EventBase],
+        ephemeral: Optional[JsonDict] = None,
+    ) -> AppServiceTransaction:
         """Atomically creates a new transaction for this application service
         with the given list of events.
```
Comment (clokep) [inline synapse/storage/databases/main/appservice.py:185]: Should this say anything about the ephemeral events coming in? It doesn't seem to get persisted, that might be worth mentioning?

File: tests/appservice/test_scheduler.py [hunk 18]
```diff
@@ -239,21 +241,59 @@ def test_multiple_service_queues(self):
...
+
+        # Send an event and don't resolve it just yet.
+        self.queuer.enqueue_ephemeral(service, event_list_1)
+
+        self.queuer.enqueue_ephemeral(service, event_list_2)
+        self.queuer.enqueue_ephemeral(service, event_list_3)
+        self.txn_ctrl.send.assert_called_with(service, [], event_list_1)
+        self.assertEquals(1, self.txn_ctrl.send.call_count)
+        # Resolve the send event: expect the queued events to be sent
+        d.callback(service)
```
Comment (clokep) [inline tests/appservice/test_scheduler.py:297]: I'm not really sure what this is doing. How does this resolve the event?

File: synapse/storage/databases/main/receipts.py [hunk 19]
```diff
@@ -123,6 +123,15 @@ def f(txn):
             for row in rows
         }
 
+    async def get_linearized_receipts_for_all_rooms(
```
Comment (clokep) [inline synapse/storage/databases/main/receipts.py:126]: It doesn't seem like this method provides and value and `_get_linearized_receipts_for_all_rooms` could just be renamed to this?

**Review by clokep — CHANGES_REQUESTED** [review]
I'm sure I've missed some things here still since this is rather large. In the future it would probably be best to add type hints for existing code as a separate PR.

File: synapse/config/appservice.py [hunk 20]
```diff
@@ -160,6 +160,8 @@ def _load_appservice(hostname, as_info, config_filename):
     if as_info.get("ip_range_whitelist"):
         ip_range_whitelist = IPSet(as_info.get("ip_range_whitelist"))
 
+    supports_ephemeral = as_info.get("de.sorunome.msc2409.push_ephemeral", False)
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/config/appservice.py:163]: Nope. One is a configuration flag in the registration file which would be `push_ephemeral` when the MSC is merged.

File: synapse/appservice/__init__.py [hunk 21]
```diff
@@ -161,19 +132,21 @@ def _check_namespaces(self, namespaces):
...
         if regex_obj:
             return regex_obj["exclusive"]
         return False
 
-    async def _matches_user(self, event, store):
+    async def _matches_user(
+        self, event: EventBase, store: Optional["DataStore"] = None
+    ) -> bool:
         if not event:
             return False
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/appservice/__init__.py:151]: Should have been optional

File: synapse/appservice/scheduler.py [hunk 22]
```diff
@@ -96,25 +104,31 @@ class _ServiceQueuer:
...
-
         if service.id in self.requests_in_flight:
             return
 
         run_as_background_process(
             "as-sender-%s" % (service.id,), self._send_request, service
         )
 
-    async def _send_request(self, service):
+    def enqueue(self, service, event):
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/appservice/scheduler.py:123]: Yeah, good shout.

File: synapse/appservice/scheduler.py [hunk 23]
```diff
@@ -123,10 +137,11 @@ async def _send_request(self, service):
         try:
             while True:
                 events = self.queued_events.pop(service.id, [])
-                if not events:
+                ephemeral = self.queued_ephemeral.pop(service.id, None)
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/appservice/scheduler.py]: I think I was doing something quirky that didn't need to be, have removed it.

File: synapse/handlers/appservice.py (hunk 13 above)
Comment (Half-Shot) [reply synapse/handlers/appservice.py:207]: Decided to put it inside the if events for both, to save us a DB call if we didn't actually handle anything.

File: synapse/storage/databases/main/appservice.py (hunk 17 above)
Comment (Half-Shot) [reply synapse/storage/databases/main/appservice.py:185]: Have mentioned a thing.

File: tests/appservice/test_scheduler.py (hunk 18 above)
Comment (Half-Shot) [reply tests/appservice/test_scheduler.py]: Have tried to clarify

File: synapse/storage/databases/main/receipts.py (hunk 19 above)
Comment (Half-Shot) [reply synapse/storage/databases/main/receipts.py]: Renamed.

File: synapse/appservice/api.py [hunk 24]
```diff
@@ -211,15 +218,18 @@ async def push_bulk(self, service, events, txn_id=None):
...
                 "push_bulk: Missing txn ID sending events to %s", service.url
             )
-            txn_id = str(0)
-        txn_id = str(txn_id)
+            txn_id = 0
+
+        uri = service.url + ("/transactions/%s" % urllib.parse.quote(str(txn_id)))
+
+        if ephemeral:
+            body = {"events": events, "de.sorunome.msc2409.ephemeral": ephemeral}
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/appservice/api.py:227]: Have requested [here](https://github.com/matrix-org/matrix-doc/pull/2409/files#r498236243). The author appears to be a bit busy.

[the remark this answers is not in the document]
Comment (Half-Shot) [reply synapse/appservice/api.py:227]: I have got them to update the MSC.

File: synapse/config/appservice.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/config/appservice.py:163]: Yeah I see that now on the MSC! Thanks for clarifying.

File: synapse/storage/databases/main/receipts.py [hunk 25]
```diff
@@ -274,6 +274,49 @@ def f(txn):
         }
         return results
 
+    @cached(num_args=2,)
+    async def get_linearized_receipts_for_all_rooms(
+        self, to_key: int, from_key: Optional[int] = None
+    ):
```
Comment (clokep) [inline synapse/storage/databases/main/receipts.py:280]: It would be ideal to add a docstring and a return type to this (which I think is `Dict[str, JsonDict]`).

File: synapse/appservice/__init__.py [hunk 26]
```diff
@@ -276,22 +290,22 @@ def get_exclusive_user_regexes(self):
...
+    def get_groups_for_user(self, user_id: str) -> Iterable[str]:
         """Get the groups that this user is associated with by this AS
 
         Args:
-            user_id (str): The ID of the user.
+            user_id: The ID of the user.
 
         Returns:
-            iterable[str]: an iterable that yields group_id strings.
+            iterable: an iterable that yields group_id strings.
```
Comment (clokep) [inline synapse/appservice/__init__.py:300]:
```suggestion
            An iterable that yields group_id strings.
```

File: synapse/appservice/scheduler.py [hunk 27]
```diff
@@ -96,25 +104,31 @@ class _ServiceQueuer:
...
-
         if service.id in self.requests_in_flight:
             return
 
         run_as_background_process(
             "as-sender-%s" % (service.id,), self._send_request, service
         )
 
-    async def _send_request(self, service):
+    def enqueue_event(self, service, event):
```
Comment (clokep) [inline synapse/appservice/scheduler.py:123]: If we're here, no reason not to add type hints:

```suggestion
    def enqueue_event(self, service: ApplicationService, event: JsonDict):
```

File: synapse/appservice/scheduler.py [hunk 28]
```diff
@@ -158,9 +173,16 @@ def __init__(self, clock, store, as_api):
         # for UTs
         self.RECOVERER_CLASS = _Recoverer
 
-    async def send(self, service, events):
+    async def send(
+        self,
+        service: ApplicationService,
+        events: List[EventBase],
+        ephemeral: Optional[JsonDict] = None,
```
Comment (clokep) [inline synapse/appservice/scheduler.py:180]: Isn't this a `Optional[List[JsonDict]]`? `enqueue_ephemeral` seems to take a `List[JsonDict]` which I think then pops back out here via `_send_request`?

File: tests/appservice/test_scheduler.py [hunk 29]
```diff
@@ -239,21 +241,58 @@ def test_multiple_service_queues(self):
...
+        self.queuer.enqueue_ephemeral(service, event_list_1)
+        # Send more events: expect send() to NOT be called multiple times.
+        self.queuer.enqueue_ephemeral(service, event_list_2)
+        self.queuer.enqueue_ephemeral(service, event_list_3)
+        self.txn_ctrl.send.assert_called_with(service, [], event_list_1)
+        self.assertEquals(1, self.txn_ctrl.send.call_count)
+        # Resolve txn_ctrl.send
+        d.callback(service)
+        # Expect the queued events to be sent
+        self.txn_ctrl.send.assert_called_with(service, [], event_list_2 + event_list_3)
```
Comment (clokep) [inline tests/appservice/test_scheduler.py:297]: I think I'm misunderstanding some logic here, what causes `event_list_1` to get sent before the others?

File: synapse/handlers/appservice.py (hunk 11 above)
Comment (Half-Shot) [reply synapse/handlers/appservice.py:168]: I wouldn't expect the function to be too heavy on processing, but it calls some potentially expensive DB statements. I wouldn't expect this too be too horrible though, as sufficiently large homeservers should be able to use an appservice worker to offload the work.

I added some information about when it's called in the docstring. Let me know if you need more info.

File: tests/appservice/test_scheduler.py (hunk 29 above)
Comment (Half-Shot) [reply tests/appservice/test_scheduler.py:297]: - We queue event_list_1
- event_list_1 is sent (pending, because we defer the send)
- We queue event_list_2 and event_list_3, which are not sent because we've not resolved the defer.
- We resolve the deferred, processing for event_list_1 is finished.
- event_list_2 and event_list_3 are both sent because the processing was finished

File: synapse/appservice/scheduler.py (hunk 28 above)
Comment (Half-Shot) [reply synapse/appservice/scheduler.py]: I decided to make it not optional, because we use an empty list in most places.

File: tests/appservice/test_scheduler.py (hunk 29 above)
Comment (Half-Shot) [reply tests/appservice/test_scheduler.py:297]: We're aligned RE: #synapse-dev:matrix.org

File: synapse/handlers/appservice.py [hunk 30]
```diff
@@ -86,7 +87,7 @@ async def notify_interested_services(self, max_token: RoomStreamToken):
                     if not events:
                         break
 
-                    events_by_room = {}
+                    events_by_room: Dict[str, List[EventBase]] = {}
```
Comment (clokep) [inline synapse/handlers/appservice.py:90]: I don't think this will work on Python 3.5:

```suggestion
                    events_by_room = {}. # type: Dict[str, List[EventBase]]
```

Comment (Half-Shot) [reply synapse/handlers/appservice.py:90]: /me sets fire to python 3.5

File: synapse/handlers/appservice.py [hunk 31]
```diff
@@ -223,7 +222,7 @@ async def _handle_typing(self, service: ApplicationService, new_token: int):
             # For performance reasons, we don't persist the previous
             # token in the DB and instead fetch the latest typing information
             # for appservices.
-            from_key=new_token - 1,
+            from_key=int(new_token) - 1,
```
Comment (clokep) [inline synapse/handlers/appservice.py:225]: Why this change? Isn't `new_token` always an int? (According to the type hints...)

Comment (Half-Shot) [reply synapse/handlers/appservice.py]: argh, that was me trying to combat `Union` types before I realized I needed to do my type checking earlier.

**Review by clokep — APPROVED** [review]
